### PR TITLE
Implementation of Group Normalization

### PR DESCRIPTION
**Pull request by Logan-Cole-Luna**
This pull request introduces Group Normalization as an enhancement to YOLO. It is an alternative to the standard Batch Normalization used by YOLO. This implementation improves YOLO's ability to handle long-tailed datasets and a lower batch count, leading to an overall increase in YOLO accuracy. It performs notably better in handling long-tailed datasets as displayed below, making it relevant to projects that do not have an ideal, uniformly distributed dataset.

Coco128 Dataset, 1000 Epochs, no transfer weights

Batch Normalization:
Class Images Instances Box(P R mAP50 mAP50-95): 100%|██████████| 16/16 [00:01<00:00, 10.17it/s]
all 128 929 0.616 0.472 0.513 0.349
wandb: model/speed_PyTorch(ms) 11.088

Group Normalization:
YOLOv8n summary: 225 layers, 3157200 parameters, 0 gradients, 8.7 GFLOPs
Class Images Instances Box(P R mAP50 mAP50-95): 100%|██████████| 16/16 [00:01<00:00, 12.36it/s]
all 128 929 0.68 0.486 0.549 0.376
wandb: model/speed_PyTorch(ms) 14.852

Another exciting result is that Group Normalization displayed its effectiveness in handling a long-tail distribution dataset such as coco128 which is primarily focused on the person class:

Batch Norm:
Class Images Instances Box(P R mAP50 mAP50-95): 100%|██████████|
all 128 929 0.616 0.472 0.513 0.349
bicycle 3 6 0.749 0.167 0.17 0.136
car 12 46 0.299 0.0563 0.0682 0.0292
motorcycle 4 5 0.661 0.8 0.92 0.748
airplane 5 6 0.32 0.833 0.638 0.419
bus 5 7 0.623 0.714 0.694 0.606
train 3 3 0.355 1 0.913 0.62
truck 5 12 1 0.305 0.479 0.338
boat 2 6 0.175 0.667 0.184 0.1
traffic light 4 14 0.849 0.143 0.15 0.12
stop sign 2 2 1 0 0.503 0.119
bench 5 9 0.217 0.249 0.298 0.154
bird 2 16 0.933 0.688 0.905 0.57
cat 4 4 0.712 0.75 0.825 0.595
dog 9 9 0.205 0.46 0.27 0.166

Group Norm:
Class Images Instances Box(P R mAP50 mAP50-95): 100%|██████████|
bicycle 3 6 0.803 0.167 0.169 0.118
car 12 46 0.695 0.0435 0.0913 0.0522
motorcycle 4 5 0.706 0.8 0.796 0.645
airplane 5 6 0.898 1 0.995 0.7
bus 5 7 0.676 0.429 0.585 0.49
train 3 3 0.759 1 0.995 0.735
truck 5 12 0.825 0.395 0.454 0.313
boat 2 6 0.698 0.396 0.647 0.339
traffic light 4 14 0.756 0.143 0.162 0.125
stop sign 2 2 0.274 0.5 0.498 0.349
bench 5 9 0.426 0.444 0.428 0.269
bird 2 16 0.651 0.625 0.669 0.38
cat 4 4 0.897 1 0.995 0.737
dog 9 9 0.691 0.889 0.828 0.593

An MRE to replicate the results displayed above:

``` python
import torch
from ultralytics import YOLO

# Load the model, ensure it has the specification for norm_type: "group", the default currently in yolov8.yaml
# is currently norm_type: "batch" as this is the standard normalization used in YOLO
# Please be sure to change it before running to view changes
model = YOLO("yolov8n.yaml")#.load("yolov8n.pt")  # build from YAML and transfer weights

# Print the normalization type, default to 'batch' if not specified
print(f'\nNormalization type: {model.model.yaml.get("norm_type", "batch")} normalization\n')

results = model.train(data="coco128.yaml", epochs=1000, imgsz=640, batch=4")
```

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
This PR adds flexible normalization support to YOLO models, allowing users to choose between Batch Normalization and Group Normalization for improved adaptability across different training scenarios. 🧑‍🔬✨

---

### 📊 Key Changes
- **Normalization Type Option:**  
  - Introduces a `norm_type` parameter (e.g., `"batch"` or `"group"`) to model configs and core modules.
  - All major model blocks (e.g., Conv, Bottleneck, C2f, Detect, Segment, etc.) now accept and use this parameter.
- **Config Updates:**  
  - YOLOv8 and YOLOv10n YAML configs updated to specify `norm_type: "batch"` by default.
- **GroupNorm Support:**  
  - Group Normalization is now available as an alternative to BatchNorm in convolutional and other layers.
  - Proper initialization and handling of GroupNorm layers added.
- **Model Parsing & Fusion:**  
  - Model parser and layer fusion logic updated to handle both BatchNorm and GroupNorm correctly.
  - Fusion (combining Conv and Norm layers for inference speed) skips GroupNorm, which is not fusable.
- **Head Modules Updated:**  
  - Detection, segmentation, pose, and classification heads now support the new normalization parameter.
- **Backward Compatibility:**  
  - Default behavior remains unchanged for existing models (BatchNorm is still default).

---

### 🎯 Purpose & Impact
- **Greater Flexibility:**  
  - Users can now select the normalization method best suited for their hardware or dataset, improving training stability and performance in various scenarios.
- **Better Support for Small Batches:**  
  - GroupNorm is particularly useful when training with small batch sizes, where BatchNorm may be less effective.
- **Easier Customization:**  
  - Advanced users and researchers can experiment with different normalization strategies without modifying core code.
- **Maintains Compatibility:**  
  - Existing workflows and models continue to work as before, with BatchNorm as the default.

---

**In short:**  
This update makes YOLO models more adaptable and robust for diverse training needs, especially in research and edge cases! 🚀🧩

### CODE REVIEW

**Comment by Logan-Cole-Luna** [comment]
I have read the CLA Document and I sign the CLA

**Comment by Logan-Cole-Luna** [comment]
@glenn-jocher 
Hi Glenn, here is the completely updated version of Group Norm that we talked about a few days ago. Please let me know if you need anything else and I look forward to your response!

**Comment by glenn-jocher** [comment]
Hi @Logan-Cole-Luna,

Thank you for the updated implementation of Group Normalization! We appreciate your efforts in enhancing YOLO's flexibility and performance. 

To proceed, could you please provide a minimum reproducible example (MRE) that demonstrates the changes and their impact? This will help us thoroughly test and validate the new feature. You can refer to our [Minimum Reproducible Example Guide](https://docs.ultralytics.com/help/minimum_reproducible_example) for more details on creating an effective MRE.

Additionally, please ensure that your implementation has been tested with the latest versions of the Ultralytics packages to confirm that the issue persists and the new feature integrates seamlessly.

Looking forward to your response and further collaboration on this enhancement! 🚀

**Comment by Logan-Cole-Luna** [comment]
@glenn-jocher 
Hi again,

Below is an MRE for the changes and everything is up to date with the latest version of Ultralytics.

```python
import torch
from ultralytics import YOLO

# Load the model, ensure it has the specification for norm_type: "group", the default currently in yolov8.yaml
# is currently norm_type: "batch" as this is the standard normalization used in YOLO
# Please be sure to change it before running to view changes
model = YOLO("yolov8n.yaml")#.load("yolov8n.pt")  # build from YAML and transfer weights

# Print the normalization type, default to 'batch' if not specified
print(f'\nNormalization type: {model.model.yaml.get("norm_type", "batch")} normalization\n')

results = model.train(data="coco128.yaml", epochs=1000, imgsz=640, batch=4")
```

**Comment by glenn-jocher** [comment]
Hi @Logan-Cole-Luna-Luna,

Thank you for providing the Minimum Reproducible Example (MRE) and confirming that everything is up to date with the latest version of Ultralytics. Your detailed example is very helpful!

We'll proceed with testing the Group Normalization implementation to ensure it integrates seamlessly and enhances the performance as expected. If we encounter any issues or have further questions, we'll reach out to you.

In the meantime, if you have any additional insights or data from your experiments that you think might be useful, please feel free to share them here.

Thanks again for your contribution! 🚀

**Comment by Logan-Cole-Luna** [comment]
Hello @glenn-jocher,

I wanted to follow up on whether you plan to incorporate Group Norm.

In my workplace, models trained with this Group Norm branch have consistently outperformed those using Batch Norm, without any increase in inference time. The results have been exceptional, and I believe this approach could significantly benefit other projects, especially those handling large datasets and struggling with performance issues using Batch Norm.

While I'm unable to share specific training details and results, the effectiveness of Group Norm in my experience has been very good.

Looking forward to your response!

**Comment by glenn-jocher** [comment]
Hello,

Thank you for sharing your positive experience with Group Normalization. We appreciate your feedback and are currently evaluating the implementation to ensure it integrates well with our framework. Your insights on its performance benefits are valuable, and we aim to provide the best tools for our users. We'll keep the community updated on our progress and any decisions regarding its incorporation.

**Comment by Logan-Cole-Luna** [comment]
@glenn-jocher 
Thank you for the update! I look forward to hearing your thoughts. Please feel free to let me know if any changes are needed on my end to make sure the code fits seamlessly into the repository. I'm happy to assist with any adjustments necessary.

**Comment by glenn-jocher** [comment]
Thank you for your patience and contribution. We are currently reviewing the implementation and will reach out if any changes are needed. Your willingness to assist is greatly appreciated.
